File: java/383.java
// 383. Ransom Note
//
// Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
//
// Each letter in magazine can only be used once in ransomNote.
//
//
//
// Example 1:
//
// Input: ransomNote = "a", magazine = "b"
// Output: false
// Example 2:
//
// Input: ransomNote = "aa", magazine = "ab"
// Output: false
// Example 3:
//
// Input: ransomNote = "aa", magazine = "aab"
// Output: true
//
//
// Constraints:
//
// 1 <= ransomNote.length, magazine.length <= 105
// ransomNote and magazine consist of lowercase English letters.
//
// Runtime 4 ms Beats 67.54%
// Memory 44.3 MB Beats 5.50%
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] magazineCounts = new int[26];
        int[] ransomCounts = new int[26];
        for (int i = 0; i < ransomNote.length(); i++) {
            char c = ransomNote.charAt(i);
            ransomCounts[c - 'a']++;
        }
        for (int i = 0; i < magazine.length(); i++) {
            char c = magazine.charAt(i);
            magazineCounts[c - 'a']++;
        }
        for (int i = 0; i < 26; i++) {
            if (ransomCounts[i] > magazineCounts[i]) return false;
        }
        return true;
    }
}

// Runtime 4 ms Beats 67.54%
// Memory 43.7 MB Beats 5.50%
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote.length() > magazine.length()) return false;
        int[] magazineCounts = new int[26];
        for (int i = 0; i < magazine.length(); i++) {
            char c = magazine.charAt(i);
            magazineCounts[c - 'a']++;
        }
        for (int i = 0; i < ransomNote.length(); i++) {
            char c = ransomNote.charAt(i);
            if (magazineCounts[c - 'a'] == 0) return false;
            magazineCounts[c - 'a']--;
        }
        return true;
    }
}

File: go/383.go
// 383. Ransom Note

// Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

// Each letter in magazine can only be used once in ransomNote.

 

// Example 1:

// Input: ransomNote = "a", magazine = "b"
// Output: false
// Example 2:

// Input: ransomNote = "aa", magazine = "ab"
// Output: false
// Example 3:

// Input: ransomNote = "aa", magazine = "aab"
// Output: true
 

// Constraints:

// 1 <= ransomNote.length, magazine.length <= 105
// ransomNote and magazine consist of lowercase English letters.

// Runtime 10 ms Beats 45.17%
// Memory 5.85 MB Beats 51.50%
func canConstruct(ransomNote string, magazine string) bool {
    counts := make(map[rune]int)
    for _, char := range magazine {
        counts[char]++
    }
    for _, char := range ransomNote {
        if counts[char] > 0 {
            counts[char]--
        } else {
            return false
        }
    }
    return true
}


// Runtime 0 ms Beats 100.00%
// Memory 5.69 MB Beats 99.42%
func canConstruct(ransomNote string, magazine string) bool {
    var counts [26]int
    for _, char := range magazine {
        counts[char - rune('a')]++
    }
    for _, char := range ransomNote {
        if counts[char - rune('a')] > 0 {
            counts[char - rune('a')]--
        } else {
            return false
        }
    }
    return true
}